rust/src/settings.rs

```diff
@@ -0,0 +1,224 @@
+//! Module for loading and validating coordinator settings.
+//!
+//! Values defined in the configuration file can be overridden by environment variables.
+use crate::mask::{BoundType, DataType, GroupType, MaskConfig, ModelType};
+use config::{Config, ConfigError, Environment};
+use std::path::PathBuf;
+use thiserror::Error;
+use validator::{Validate, ValidationErrors};
+
+#[derive(Error, Debug)]
+pub enum SettingsError {
+    #[error("configuration loading failed: {0}")]
+    Loading(#[from] ConfigError),
+    #[error("validation failed: {0}")]
+    Validation(#[from] ValidationErrors),
+}
+
+#[derive(Debug, Validate, Deserialize, Clone, Copy)]
+#[doc(hidden)]
+pub struct Settings {
+    #[validate]
+    pub api: ApiSettings,
+    #[validate]
+    pub pet: PetSettings,
+    pub mask: MaskSettings,
+}
+
+#[derive(Debug, Validate, Deserialize, Clone, Copy)]
+/// PET protocol settings
+pub struct PetSettings {
```

Comment:
what do you think of adding an additional struct-level validation for `PetSettings`? it would allow checking for the corner cases mentioned last week in the slack channel.

it should be enough to add this attribute to the `PetSettings`:
```
#[validate(schema(function = "validate_fractions"))]
```
where the function `"validate_fractions"` will be
```rust
fn validate_fractions(s: PetSettings) -> Result<(), ValidationError> {
    if 0. < s.sum
        && s.sum < 1.
        && 0. < s.update
        && s.update < 1.
        && 0 < s.sum + s.update - s.sum*s.update
        && s.sum + s.update - s.sum*s.update < 1
    {
        Ok(())
    } else {
        Err(ValidationError::new("starvation"))
    }
}
```
(i'm not sure on the input arguments, the documentation is quite lacking.)

Reply:
Good point👍

I will add this

Reply:
Seems to work!
If we add this validation we can actually remove the validation `#[validate(range(min = 0, max = 1.0))]` on the `sum` and `update` fields right?

Reply:
sure, the validation on the fields is now subsumed by the validation of the struct. we should still document the constraints somewhere.